crates/flipperzero/src/toolbox/md5.rs

```diff
@@ -0,0 +1,126 @@
+use core::fmt;
+
+use digest::{
+    block_buffer::Eager,
+    core_api::{Buffer, BufferKindUser, CoreWrapper, FixedOutputCore, UpdateCore},
+    crypto_common::{AlgorithmName, Block, BlockSizeUser},
+    typenum::{Unsigned, U16, U64},
+    HashMarker, Output, OutputSizeUser, Reset,
+};
+use flipperzero_sys as sys;
+
+/// The [MD5 hash function][1].
+///
+/// Equivalent to [`md5::Md5`].
+///
+/// ## ⚠️ Security Warning
+///
+/// This type is provided for the purposes of legacy interoperability with protocols and
+/// systems which mandate the use of MD5.
+///
+/// However, MD5 is [cryptographically broken and unsuitable for further use][2].
+///
+/// Collision attacks against MD5 are both practical and trivial, and
+/// [theoretical attacks against MD5's preimage resistance have been found][3].
+///
+/// [RFC6151][4] advises no new IETF protocols can be designed using MD5-based
+/// constructions, including HMAC-MD5.
+///
+/// [1]: https://en.wikipedia.org/wiki/MD5
+/// [2]: https://www.kb.cert.org/vuls/id/836068
+/// [3]: https://dl.acm.org/citation.cfm?id=1724151
+/// [4]: https://tools.ietf.org/html/rfc6151
+pub type Md5 = CoreWrapper<Md5Core>;
+
+/// Core MD5 hasher.
+pub struct Md5Core {
+    state: sys::md5_context,
+}
+
+impl HashMarker for Md5Core {}
+
+impl BlockSizeUser for Md5Core {
+    type BlockSize = U64;
+}
+
+impl BufferKindUser for Md5Core {
+    type BufferKind = Eager;
+}
+
+impl OutputSizeUser for Md5Core {
+    type OutputSize = U16;
+}
+
+impl Default for Md5Core {
+    #[inline]
+    fn default() -> Self {
+        let mut state = sys::md5_context {
+            total: [0; 2],
+            state: [0; 4],
+            buffer: [0; 64],
+        };
+        unsafe { sys::md5_starts(&mut state) };
+        Self { state }
+    }
+}
+
+impl UpdateCore for Md5Core {
+    #[inline]
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        for block in blocks {
+            self.state.total[0] += Self::BlockSize::U32;
```

Comment:
I'm honestly a bit lost by this syntax (the docs are pretty opaque). `Self::BlockSize` is `typenum::U64`, so that makes `Self::BlockSize::U32`  equivalent to `typenum::U64::U32`. Is this just a compile-time type representing `64u32`?

If so, is `Self::BlockSize::to_u32()` equivalent (though an inline function vs. constant)?

Reply:
> Is this just a compile-time type representing `64u32`?

Yes. Rust lacked sufficient const generics to perform arithmetic in the type system using integers (and even for very recent Rust versions it still lacks a few things that some use cases need). `typenum` is a workaround that uses types instead,.

> If so, is `Self::BlockSize::to_u32()` equivalent (though an inline function vs. constant)?

Yep. The associated constant is provided for use in `const fn`s, and I use it because it's shorter, but I can use the function instead if you prefer (which adds a bit of indirection, but the compiler _should_ inline it).

Reply:
Nice. Thanks for explaining. I think it's fine to keep it the current way.

Maybe throw in a `// 64u32` next to the line just to nudge people in the right direction.